tools/nemo_forced_aligner/align.py

```diff
@@ -0,0 +1,287 @@
+    else:
+        if not is_entry_in_all_lines(cfg.manifest_filepath, "text"):
+            raise RuntimeError(
+                "At least one line in cfg.manifest_filepath does not contain a 'text' entry. "
+                "NFA requires all lines to contain a 'text' entry when cfg.align_using_pred_text=True."
+            )
+
+    # init devices
+    if cfg.transcribe_device is None:
+        transcribe_device = torch.device("cuda" if torch.cuda.is_available else "cpu")
+    else:
+        transcribe_device = torch.device(cfg.transcribe_device)
+    logging.info(f"Device to be used for transcription step (`transcribe_device`) is {transcribe_device}")
+
+    if cfg.viterbi_device is None:
+        viterbi_device = torch.device("cuda" if torch.cuda.is_available else "cpu")
+    else:
+        viterbi_device = torch.device(cfg.viterbi_device)
+    logging.info(f"Device to be used for viterbi step (`viterbi_device`) is {viterbi_device}")
+
+    if transcribe_device.type == 'cuda' or viterbi_device.type == 'cuda':
+        logging.warning(
+            'One or both of transcribe_device and viterbi_device are GPUs. If you run into OOM errors '
+            'it may help to change both devices to be the CPU.'
+        )
+
+    # load model
+    model, _ = setup_model(cfg, transcribe_device)
+
+    if not isinstance(model, EncDecCTCModel):
+        raise NotImplementedError(
+            f"Model {cfg.model_name} is not an instance of NeMo EncDecCTCModel."
+            " Currently only instances of EncDecCTCModels are supported"
+        )
+
+    audio_sr = get_audio_sr(cfg.manifest_filepath)
+    logging.info(
+        f"Detected audio sampling rate {audio_sr}Hz in first audio in manifest at {cfg.manifest_filepath}. "
+        "Will assume all audios in manifest have this sampling rate. Sampling rate will be used to determine "
+        "timestamps in output CTM."
+    )
+
+    if cfg.minimum_timestamp_duration > 0:
+        logging.warning(
+            f"cfg.minimum_timestamp_duration has been set to {cfg.minimum_timestamp_duration} seconds. "
+            "This may cause the alignments for some tokens/words/additional segments to be overlapping."
+        )
+
+    # get start and end line IDs of batches
+    starts, ends = get_batch_starts_ends(cfg.manifest_filepath, cfg.batch_size)
+
+    if cfg.align_using_pred_text:
+        # record pred_texts to save them in the new manifest at the end of this script
+        pred_text_all_lines = []
+    else:
+        pred_text_all_lines = None
+
+    # get alignment and save in CTM batch-by-batch
+    for start, end in zip(starts, ends):
+        manifest_lines_batch = get_manifest_lines_batch(cfg.manifest_filepath, start, end)
+
+        (
+            log_probs_batch,
+            y_batch,
+            T_batch,
+            U_batch,
+            token_info_batch,
+            word_info_batch,
+            segment_info_batch,
+            pred_text_batch,
+        ) = get_batch_tensors_and_boundary_info(
```

Comment:
No need to do this now, but I **really** suggest creating a dataclass that contains all these values as the return type here.